From fe77e2277ce4f13ca93fab8eb450e7c480e71161 Mon Sep 17 00:00:00 2001
From: Oliver Bestmann <oliver.bestmann@googlemail.com>
Date: Tue, 16 Dec 2025 11:59:22 +0100
Subject: [PATCH] drm: apple: set timestamps for 120hz

The dcp does not seem to care much about the values in ts1, ts2 and ts3, as long
as they are non zero. This commit fills the timestamp with a dummy value
of 120 if a refresh-rate of 120hz is selected.
This is enough to get a refresh rate of 120hz.

MacOS also sets flags1 and flags2. I have no idea what exactly those values
indicate, but I did do not need to set any of them to get 120hz.

Signed-off-by: Oliver Bestmann <oliver.bestmann@googlemail.com>
---
 drivers/gpu/drm/apple/dcp-internal.h   |  1 +
 drivers/gpu/drm/apple/iomfb_template.c | 13 +++++++++++++
 drivers/gpu/drm/apple/iomfb_template.h |  9 ++++++++-
 drivers/gpu/drm/apple/parser.c         | 14 +-------------
 drivers/gpu/drm/apple/parser.h         |  1 +
 5 files changed, 24 insertions(+), 14 deletions(-)

diff --git a/drivers/gpu/drm/apple/dcp-internal.h b/drivers/gpu/drm/apple/dcp-internal.h
index 2c31d2a8cef09d..3b6a14339cddef 100644
--- a/drivers/gpu/drm/apple/dcp-internal.h
+++ b/drivers/gpu/drm/apple/dcp-internal.h
@@ -184,6 +184,7 @@ struct apple_dcp {
 	/* Current display mode */
 	bool during_modeset;
 	bool valid_mode;
+	s64 precise_sync_rate;
 	struct dcp_set_digital_out_mode_req mode;
 
 	/* completion for active turning true */
diff --git a/drivers/gpu/drm/apple/iomfb_template.c b/drivers/gpu/drm/apple/iomfb_template.c
index 0a1b9495128562..1898396c276381 100644
--- a/drivers/gpu/drm/apple/iomfb_template.c
+++ b/drivers/gpu/drm/apple/iomfb_template.c
@@ -1219,6 +1219,9 @@ int DCP_FW_NAME(iomfb_modeset)(struct apple_dcp *dcp,
 		.timing_mode_id = mode->timing_mode_id
 	};
 
+	/* keep around to fake timestamps in dcp_swap_submit */
+	dcp->precise_sync_rate = mode->precise_sync_rate;
+
 	cookie = kzalloc(sizeof(*cookie), GFP_KERNEL);
 	if (!cookie) {
 		return -ENOMEM;
@@ -1408,6 +1411,16 @@ void DCP_FW_NAME(iomfb_flush)(struct apple_dcp *dcp, struct drm_crtc *crtc, stru
 		req->clear = 1;
 	}
 
+	if (has_surface && (dcp->precise_sync_rate >> 16) == 120) {
+		/*
+		 * Fake timstamps to get 120hz refresh rate. It looks
+		 * like the actual value does not matter, as long  as it is non zero.
+		 */
+		req->swap.ts1 = 120;
+		req->swap.ts2 = 120;
+		req->swap.ts3 = 120;
+	}
+
 	/* These fields should be set together */
 	req->swap.swap_completed = req->swap.swap_enabled;
 
diff --git a/drivers/gpu/drm/apple/iomfb_template.h b/drivers/gpu/drm/apple/iomfb_template.h
index f446a4d8f38b90..878ffdf9e644cd 100644
--- a/drivers/gpu/drm/apple/iomfb_template.h
+++ b/drivers/gpu/drm/apple/iomfb_template.h
@@ -18,7 +18,14 @@
 struct DCP_FW_NAME(dcp_swap) {
 	u64 ts1;
 	u64 ts2;
-	u64 unk_10[6];
+
+	u64 unk_10;
+	u64 unk_18;
+	u64 ts64_unk;
+	u64 unk_28;
+	u64 ts3;
+	u64 unk_38;
+
 	u64 flags1;
 	u64 flags2;
 
diff --git a/drivers/gpu/drm/apple/parser.c b/drivers/gpu/drm/apple/parser.c
index 700a31883eac8e..e44b4e17758a55 100644
--- a/drivers/gpu/drm/apple/parser.c
+++ b/drivers/gpu/drm/apple/parser.c
@@ -499,19 +499,6 @@ static int parse_mode(struct dcp_parse_ctx *handle,
 	if (is_virtual)
 		return -EINVAL;
 
-	/*
-	 * HACK:
-	 * Ignore the 120 Hz mode on j314/j316 (identified by resolution).
-	 * DCP limits normal swaps to 60 Hz anyway and the 120 Hz mode might
-	 * cause choppiness with X11.
-	 * Just downscoring it and thus making the 60 Hz mode the preferred mode
-	 * seems not enough for some user space.
-	 */
-	if (vert.precise_sync_rate >> 16 == 120 &&
-	    ((horiz.active == 3024 && vert.active == 1964) ||
-	     (horiz.active == 3456 && vert.active == 2234)))
-		return -EINVAL;
-
 	vert.active -= notch_height;
 	vert.sync_width += notch_height;
 
@@ -539,6 +526,7 @@ static int parse_mode(struct dcp_parse_ctx *handle,
 
 	out->timing_mode_id = id;
 	out->color_mode_id = best_color_mode;
+	out->precise_sync_rate = vert.precise_sync_rate;
 
 	trace_iomfb_timing_mode(handle->dcp, id, *score, horiz.active,
 				vert.active, vert.precise_sync_rate,
diff --git a/drivers/gpu/drm/apple/parser.h b/drivers/gpu/drm/apple/parser.h
index 2f52e063bbd426..a8add87a3d438c 100644
--- a/drivers/gpu/drm/apple/parser.h
+++ b/drivers/gpu/drm/apple/parser.h
@@ -87,6 +87,7 @@ struct dcp_display_mode {
 	struct drm_display_mode mode;
 	u32 color_mode_id;
 	u32 timing_mode_id;
+	s64 precise_sync_rate;
 	struct dcp_color_mode sdr_rgb;
 	struct dcp_color_mode sdr_444;
 	struct dcp_color_mode sdr;
